File: 2024.1/semana07/f-leet-733-flood-fill.exs
defmodule Solution do
  defp get_or_nil(image, x, y) do
    cond do
      x < 0 or y < 0 or x >= tuple_size(image) or y >= tuple_size(elem(image, 0)) -> nil
      true -> image |> elem(x) |> elem(y)
    end
  end

  @spec flood_fill(image :: [[integer]], sr :: integer, sc :: integer, color :: integer) :: [[integer]]
  def flood_fill(image, sr, sc, color) do
    image
      |> Enum.map(&List.to_tuple/1)
      |> List.to_tuple()
      |> (&fill(&1, sr, sc, color, get_or_nil(&1, sr, sc))).()
      |> Tuple.to_list()
      |> Enum.map(&Tuple.to_list/1)
  end

  defp fill(image, _x, _y, color, oldcolor) when color == oldcolor, do: image

  @movs [{0, 1}, {0, -1}, {1, 0}, {-1, 0}]
  defp fill(image, x, y, color, oldcolor) do 
    if get_or_nil(image, x, y) == oldcolor do
        image = image
        |> elem(x)
        |> put_elem(y, color)
        |> (&put_elem(image, x, &1)).()

        @movs
          |> Enum.reduce(image, fn {ax, ay}, img -> fill(img, x + ax, y + ay, color, oldcolor) end)
    else
        image
    end
  end
end

Solution.flood_fill([[1,1,1], [1,1,0], [1,0,1]], 1, 1, 2)
  |> IO.inspect(label: "final")

File: 2024.1/semana07/f-leet-733-flood-fill.cpp
#include <bits/stdc++.h>

using namespace std;

const int mx[] = {1, -1, 0, 0};
const int my[] = {0, 0, 1, -1};
class Solution {
  public:
    void transverse(vector<vector<int>> &grid, int limitX, int limitY, int x,
                    int y, int orig, int target) {
        if (grid[x][y] != orig || grid[x][y] == target)
            return;
        grid[x][y] = target;

        for (int i = 0; i < 4; i++) {
            int cx = x + mx[i];
            int cy = y + my[i];
            if (cx >= 0 && cx < limitX && cy >= 0 && cy < limitY)
                transverse(grid, limitX, limitY, cx, cy, orig, target);
        }
    }
    vector<vector<int>> floodFill(vector<vector<int>> &image, int sr, int sc,
                                  int color) {
        transverse(image, image.size(), image[0].size(), sr, sc, image[sr][sc],
                   color);

        return image;
    }
};
